vortex-array/src/compute/scalar_subtract.rs

```diff
@@ -0,0 +1,42 @@
+use vortex_dtype::DType;
+use vortex_error::{vortex_err, VortexResult};
+
+use crate::scalar::Scalar;
+use crate::{Array, ArrayDType, Flattened, OwnedArray};
+
+pub trait ScalarSubtractFn {
+    fn scalar_subtract(&self, to_subtract: &Scalar) -> VortexResult<OwnedArray>;
+}
+
+pub fn scalar_subtract<T: Into<Scalar>>(array: &Array, to_subtract: T) -> VortexResult<OwnedArray> {
+    let to_subtract = to_subtract.into().cast(array.dtype())?;
+
+    if let Some(subtraction_result) =
+        array.with_dyn(|c| c.scalar_subtract().map(|t| t.scalar_subtract(&to_subtract)))
+    {
+        subtraction_result
+    } else {
+        // if subtraction is not implemented for the given array type, but the array has a numeric
+        // DType, we can flatten the array and apply subtraction to the flattened primitive array
+        let result = match array.dtype() {
+            DType::Int(..) | DType::Decimal(..) | DType::Float(..) => {
+                let array = array.clone();
+                let result = array.flatten()?;
```

Comment:
For Int and Float, you know they must flatten into a primitive. So you can use flatten_primitive.